gcp_variant_transforms/beam_io/bgzf_io.py

```diff
@@ -0,0 +1,176 @@
+# Copyright 2019 Google Inc.  All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Splits the BGZF file."""
+
+import struct
+from collections import namedtuple
+from typing import Iterable, List, Tuple  # pylint: disable=unused-import
+
+from apache_beam.io import filesystem
+from apache_beam.io import filesystems
+
+# `Block` provides real file offset of the start and end of a gzip block. Given
+# it, one can directly seek to the start of the gzip block and decompress it.
+Block = namedtuple('Block', ['start', 'end'])
+
+_INDEX_FILE_APPENDIX = '.tbi'
+
+# `MAX_BLOCK_SIZE` defines the maximum amount of data included in one `Block`.
+MAX_BLOCK_SIZE = 1 * 1024 * 1024
+# It beyonds the normal bin range (1-37449), and contains other meta data rather
+# than file offsets of GZIP blocks.
+_RESERVED_BIN_ID = 37450
+
+
+def split_bgzf(file_path):
+  # type: (str) -> Iterable[Tuple[str, Block]]
+  """Splits BGZF to multiple blocks.
+
+  It reads the `Block` (GZIP block offsets) from the corresponding index file.
+  Given (`file_path`, `Block`), one is able to decompress the data within
+  `Block`. It further:
+  - Removes invalid Blocks.
+  - Fills in the gap (by inserting a new Block) if there are gaps between two
+    adjacent Blocks to make sure it has full coverage of the variant data.
+  - Merges Blocks. It eliminates overlapping Blocks, and merges adjacent Blocks
+    if their combined size is smaller than `MAX_BLOCK_SIZE`.
+
+  Args:
+    file_path: The BGZF path.
+
+  Yields:
+    (file_path, Block).
+  """
+  blocks = _get_block_offsets(_get_tbi_file(file_path))
+  blocks = _remove_invalid_blocks(blocks)
+  blocks = _fill_in_gap(blocks)
+  blocks = _merge_blocks(blocks)
+
+  for block in blocks:
+    yield (file_path, block)
+
+
+def exists_tbi_file(file_path):
+  return filesystems.FileSystems.exists(_get_tbi_file(file_path))
+
+
+def _get_tbi_file(file_path):
+  return file_path + _INDEX_FILE_APPENDIX
```

Comment:
For BAMs, we often have .bam.bai _or_ .bai.  Is this not true in this case as well?